pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,282 @@
+	for _, pod := range podsToPreempt {
+		status := v1.PodStatus{
+			Phase:   v1.PodFailed,
+			Message: message,
+			Reason:  reason,
+		}
+		// record that we are evicting the pod
+		c.recorder.Eventf(pod, v1.EventTypeWarning, reason, message)
+		// this is a blocking call and should only return when the pod and its containers are killed.
+		err := c.killPodFunc(pod, status, nil)
+		if err != nil {
+			return fmt.Errorf("preemption: pod %s failed to evict %v", format.Pod(pod), err)
+		}
+		glog.Infof("preemption: pod %s evicted successfully", format.Pod(pod))
+	}
+	return nil
+}
+
+// getPodsToPreempt returns a list of pods that could be preempted to free requests >= requirements
+func getPodsToPreempt(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	bestEffortPods := filterPods(pods, func(p *v1.Pod) bool { return qos.GetPodQOS(p) == v1.PodQOSBestEffort })
+	burstablePods := filterPods(pods, func(p *v1.Pod) bool { return qos.GetPodQOS(p) == v1.PodQOSBurstable })
+	guaranteedPods := filterPods(pods, func(p *v1.Pod) bool { return qos.GetPodQOS(p) == v1.PodQOSGuaranteed })
+
+	// We only evict guaranteed pods if we "have to".  This implies that guaranteed pods only need to fulfil requirements
+	// which are not fulfilled by all other burstable or best effort pods.
+	guarateedToEvict, err := getPodsToPreemptByDistance(guaranteedPods, requirements.subtract(append(bestEffortPods, burstablePods...)...))
+	if err != nil {
+		return nil, err
+	}
+	// Once we know the guaranteed pods being evicted, we can subtract their requirements from what we need to meet with burstable pods
+	// In similar fashion to guaranteed pods, we only need to consider requirements not met by all best effort pods.
+	burstableToEvict, err := getPodsToPreemptByDistance(burstablePods, requirements.subtract(append(bestEffortPods, guarateedToEvict...)...))
+	if err != nil {
+		return nil, err
+	}
+	bestEffortToEvict, err := getPodsToPreemptByDistance(bestEffortPods, requirements.subtract(append(burstableToEvict, guarateedToEvict...)...))
+	if err != nil {
+		return nil, err
+	}
+	return append(append(bestEffortToEvict, burstableToEvict...), guarateedToEvict...), nil
+}
+
+// finds the pods that have pod requests >= admission requirements.
+// Chooses pods that minimize "distance" to the requirements.
+// If more than one pod exists that fulfills the remaining requirements,
+// it chooses the pod that has the "smaller resource request"
+func getPodsToPreemptByDistance(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	podsToEvict := []*v1.Pod{}
+	// evict pods by shortest distance from remaining requirements, updating requirements every round.
+	for len(requirements) > 0 {
+		if len(pods) == 0 {
+			return nil, fmt.Errorf("no set of running pods found to reclaim resources: %v", requirements.toString())
```

Comment:
If there are no burstable or besteffort pods, then the first call 
`guarateedToEvict, err := getPodsToPreemptByDistance(guaranteedPods, requirements.subtract(append(bestEffortPods, burstablePods...)...))`
will find guaranteed pods that fulfill the full set of requirements.  Then, when "subtracting" those guaranteed pods from the remaining resources, we will not have any remaining resource requirements, and these functions will return immediately.